### DISCUSSION TITLE
customDatasources and plain format

### DISCUSSION THREAD

**Discussion by faenschi** [renovatebot/renovate#29410] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

Hey everyone,

I've configured a custom data source to update a digest value:

```json
"customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["*.yaml"],
      "matchStrings": [
        "sha256sum:\\s(?<currentDigest>.*)",
        "url:\\s(?<depName>.*)"
      ],
      "datasourceTemplate": "custom.sha256",
      "currentValueTemplate": "1"
    }
  ],
  "customDatasources": {
    "sha256": {
      "defaultRegistryUrlTemplate": "{{ packageName }}.sha256",
      "format": "plain",
      "transformTemplates": [
        "{\"releases\":[{\"version\": \"1\", \"digest\": <HOW TO TRANSFORM HERE >}]}"
      ]
    }
  }
```

The URL returns a plain response that only includes the sha256. How should I define the `transformTemplates` value so that my digest is replaced correctly?

regards angela

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by secustor** [comment]
You have to rename the field `version` to `digest` in initial structure described at https://docs.renovatebot.com/modules/datasource/custom/#plain

**Reply by faenschi** [reply]
You mean like that?

```
"transformTemplates": [
        "{\"releases\":[{\"version\": \"1\", \"digest\": $.releases[0].version}]}"
      ]
```

does not work, and also the log is not very helpful :-(

`{"name":"renovate","hostname":"angela-tln9b","pid":10,"level":20,"logContext":"8ZYnzgUtht9OoA2fzZNTQ","repository":"stempfela/renovate-playground","packageName":"sha256sum","currentValue":"1","datasource":"custom.sha256","newValue":"1","msg":"Could not determine new digest for update.","time":"2024-06-04T08:31:40.907Z","v":0}`

**Comment by faenschi** [comment]
Found the issue: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/custom/formats/plain.ts#L23

There is a check for the `content-type text`, but in my case it's not text, its `Content-Type: application/x-checksum` .

I also compared the fetchers, not every fetcher has this validation implemented. Would it make sense to remove this check or to add the possibility to configure it?

**Reply by secustor** [reply]
@rarkins  WDYT, this has been added mostly to protect the App from downloading binary files and creating that way a cost explosion. 

The other types have this checks in the upstream function and if one does not have it, it is an oversight on our side.

**Reply by marsom** [reply]
i found a solution for this problem, a plain text file with just one line is also a valid yaml file

therfore this here works:

```
"customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["*.yaml"],
      "matchStrings": [
        "zipArtifact:[\\S\\s]*sha256sum:[\\s](?<currentDigest>[a-z0-9]+)[\\S\\s]*[^#]url:[\\s](?<depName>.+)"
      ],
      "datasourceTemplate": "custom.sha256",
      "currentValueTemplate": "1"
    }
  ],
  "customDatasources": {
    "sha256": {
      "defaultRegistryUrlTemplate": "{{ packageName }}.sha256",
      "format": "yaml",
      "transformTemplates": [
        "{\"releases\":[{\"version\": \"1\", \"digest\": $}]}"
      ]
    }
  }
```

this work because the the yaml implenation does a http.get and doest not check any headers.

**Reply by rarkins** [reply]
I think we can probably drop this restriction

**Comment by secustor** [comment, marked as answer]
Has been changed in https://github.com/renovatebot/renovate/pull/29549
